src/engine/GroupBy.cpp

```diff
@@ -389,6 +407,154 @@ ProtoResult GroupBy::computeResult([[maybe_unused]] bool requestLaziness) {
+template <int COLS>
+size_t GroupBy::searchBlockBoundaries(
+    const std::invocable<size_t, size_t> auto& onBlockChange,
+    const IdTableView<COLS>& idTable,
+    std::vector<std::pair<size_t, Id>>& currentGroupBlock) const {
+  size_t blockStart = 0;
+
+  for (size_t pos = 0; pos < idTable.size(); pos++) {
+    checkCancellation();
+    bool rowMatchesCurrentBlock =
+        std::ranges::all_of(currentGroupBlock, [&](const auto& colIdxAndValue) {
+          return idTable(pos, colIdxAndValue.first) == colIdxAndValue.second;
+        });
+    if (!rowMatchesCurrentBlock) {
+      onBlockChange(blockStart, pos);
+      // setup for processing the next block
+      blockStart = pos;
+      for (auto& [colIdx, value] : currentGroupBlock) {
+        value = idTable(pos, colIdx);
+      }
+    }
+  }
+  return blockStart;
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_GROUP_COLUMNS, size_t OUT_WIDTH>
+cppcoro::generator<IdTable> GroupBy::computeResultLazily(
+    std::shared_ptr<const Result> subresult, std::vector<Aggregate> aggregates,
+    std::vector<HashMapAliasInformation> aggregateAliases,
+    std::vector<size_t> groupByCols, std::shared_ptr<LocalVocab> localVocab,
+    bool singleIdTable) const {
+  LazyGroupBy<NUM_GROUP_COLUMNS> lazyGroupBy{
+      *localVocab, std::move(aggregateAliases),
+      getExecutionContext()->getAllocator(), groupByCols.size()};
+
+  auto createEvaluationContextForTable = [this,
+                                          &localVocab](const IdTable& table) {
+    sparqlExpression::EvaluationContext evaluationContext{
+        *getExecutionContext(),
+        _subtree->getVariableColumns(),
+        table,
+        getExecutionContext()->getAllocator(),
+        *localVocab,
+        cancellationHandle_,
+        deadline_};
+    evaluationContext._groupedVariables = ad_utility::HashSet<Variable>{
+        _groupByVariables.begin(), _groupByVariables.end()};
+    evaluationContext._isPartOfGroupBy = true;
+    evaluationContext._variableToColumnMapPreviousResults =
+        getInternallyVisibleVariableColumns();
+    evaluationContext._previousResultsFromSameGroup.resize(getResultWidth());
+    return evaluationContext;
+  };
+
+  IdTable resultTable{getResultWidth(), getExecutionContext()->getAllocator()};
+
+  bool groupSplitAcrossTables = false;
+
+  std::vector<std::pair<size_t, Id>> currentGroupBlock;
+
+  for (IdTable& idTable : subresult->idTables()) {
+    if (idTable.empty()) {
+      continue;
+    }
+    checkCancellation();
+
+    if (currentGroupBlock.empty()) {
+      for (size_t col : groupByCols) {
+        currentGroupBlock.emplace_back(col, idTable(0, col));
+      }
+    }
+
+    sparqlExpression::EvaluationContext evaluationContext =
+        createEvaluationContextForTable(idTable);
+
+    size_t lastBlockStart = searchBlockBoundaries(
+        [&](size_t blockStart, size_t blockEnd) {
+          if (groupSplitAcrossTables) {
```

Comment:
`isFirstGroupOfBlock` would be factually incorrect, because for the first IdTable this value is initialized with false (even though it is the first group of the block)

Reply:
But as you have the correct `groupBlock` for the very first block in the first nonempty table, would initializing it `isFirstGroupOfBlock` to `true` not simply work?

Reply:
It would but it would be wasteful for a large group at the start of the first block.

Reply:
I don't quite understand, why it would be wasteful?
The assumption is, that using the `LazyGroupBy` for large blocks is efficient. 
Or am I missing something?
This is not a blocker though, but I want to understand it.